Sources/AblyAssetTrackingPublisher/State/DefaultTrackableState.swift

```diff
@@ -0,0 +1,90 @@
+//
```

Comment:
one word: testability :)
It can be mocked with ease in tests.

Protocol-oriented-programing is more flexible than OOP, below few benefits from [HERE](https://tsh.io/blog/protocol-oriented-programming-swift/):
- Types can conform to more than one protocol.
- They can also obtain default behaviours from multiple protocols.
- Unlike multiple inheritance in other programming languages, protocol extensions do not introduce any additional states.
- Protocols can be adopted by classes, structs and enums, while base classes and inheritance are restricted to class types only.
- Protocols allow retroactive modeling with extensions added to existing types.

Reply:
I am aware of protocol oriented programming. That still means the protocol should be designed well, my question above details design issues regardless of OOP or POP, and I don't think you've addressed them. 

Also, there is only one implementation of `PublisherTrackableState` anyway (I cannot find a mock implementation of it). If you did implement a mock, it would have to implement redundant implementation details (increment, reset counter etc) which would be code duplication.

Also, you should avoid mocks where possible. `DefaultTrackableState` is a very simple implementation, why mock it?

Reply:
I understand protocol oriented programming. As we both know, that still means the protocol should be designed well, my question above details design issues regardless of OOP or POP, and I don't think you've addressed them. (`DefaultX` naming, inheritance over protocols in this case). Its been a few months since I watched [the video](https://developers.apple.com/videos/play/wwdc2015/408/), but I've watched a few times over the last few years.

---

Now I am taking a look at `TestableTrackableState.swift`, and I confused. There are **setter** functions called `removeWillCall` and `removeDidCall`. They're not being used at all? IMHO It's a very unreadable class with 6 optional closures. I don't know what you're testing here.

It seems to me you are implementing a mock just so you can check the method was called. You pass a closure to this mock which will be called back when the method gets called, e.g.
```
        trackableState.unmarkMessageAsPendingDidCall {
            unmarkMessageAsPendingDidCallExpectation.fulfill()
        }
```

You are already inheriting `DefaultTrackableState` with `class TestableTrackableState: DefaultTrackableState`, so the protocol is not actually useful. Inheriting makes sense, because you don't want code duplication. I don't think adding this protocol is adding anything except more lines of code at the moment. Why not have `PublisherTrackableState` the original implementation, and `TestablePublisherTrackableState` inherit this, without protocols?

The fact a `XState` class is a protocol is already bad isn't it? It doesn't sound like a protocol. 

Also, SOLID principle: Interface Segregation Principle (ISP) is broken. This is quite a large interface (11 methods).

---

## Potential solution

Anyway, to solve this testing problem, lets think of the public API of `PublisherTrackableState`. Would anyone reasonably want to know what the state is? It is a `state` after all. We could make it less of an opaque box and allow users to check if something is in it or not? Then in your tests, you can just call this method/ field and check it. It sounds like a useful method even outside of testing.

Reply:
In general though, IMHO there is a problem with the API of this entity, regardless of its a protocol or a class.

Reply:
I've made a bit of changes here d3e9b72fe26d04ee46cb2f2af7a6b0065417a147
- I've split protocol into few smaller protocols (ISP problem)
- I've used multiple protocol inheritances
- I've removed unnecessary code from `TestableTrackableState` and added a comment

WDYT?

Reply:
It's mostly looking simpler I think. I think it's mostly good, but I think ISP is not the main problem with that interface. Would anyone realistically use those protocols separately (`typealias TrackableStateable = StateWaitable & StatePendable & StateRemovable & StateRetryable & StateSkippable`)? The main issue is that it leaks implementation details about how a class which implements it would work. IMHO That is the problem. This protocol split makes the code even more unreadable. 😢

---

Okay, tbh I did not **suggest** anything. I only asked a question:
> Isn't the trackable updates stored in PublisherSkippedLocationsState **fit nicer** as part of PublisherTrackableState? These are locations which are for the trackable. It is rightfully its state.

Let me elaborate. :) Don't assume I want you to do anything, because it should be your decision at the end of the day. I was asking what you thought, would it fit better? Don't do it if you disagree, or explain your thoughts too. It should be for discussion/ thoughts, I'm not telling you to do anything. I'm not sure if you agree or disagree with my thoughts quoted above.